pyleco/core/message.py

```diff
@@ -0,0 +1,242 @@
+
+    @header.setter
+    def header(self, value: bytes):
+        self._header = value
+        self._header_elements = None  # reset cache
+
+    @property
+    def payload(self) -> List[bytes]:
+        if self._payload == [] and self._data is not None:
+            self._payload = [serialize_data(self._data)]
+        return self._payload
+
+    @payload.setter
+    def payload(self, value: List[bytes]) -> None:
+        self._payload = value
+        self._data = None  # reset data
+
+    @property
+    def conversation_id(self) -> bytes:
+        if self._header_elements is None:
+            self._header_elements = interpret_header(self.header)
+        return self._header_elements[0]
+
+    @property
+    def message_id(self) -> bytes:
+        if self._header_elements is None:
+            self._header_elements = interpret_header(self.header)
+        return self._header_elements[1]
+
+    @property
+    def message_type(self) -> bytes:
+        if self._header_elements is None:
+            self._header_elements = interpret_header(self.header)
+        return self._header_elements[2]
+
+    @property
+    def receiver_node(self) -> bytes:
+        if self._receiver_elements is None:
+            self._receiver_elements = split_name(self.receiver)
+        return self._receiver_elements[0]
+
+    @property
+    def receiver_name(self) -> bytes:
+        if self._receiver_elements is None:
+            self._receiver_elements = split_name(self.receiver)
+        return self._receiver_elements[1]
+
+    @property
+    def sender_node(self) -> bytes:
+        if self._sender_elements is None:
+            self._sender_elements = split_name(self.sender)
+        return self._sender_elements[0]
+
+    @property
+    def sender_name(self) -> bytes:
+        if self._sender_elements is None:
+            self._sender_elements = split_name(self.sender)
+        return self._sender_elements[1]
+
+    @property
+    def data(self) -> object:
+        if self._data is None and self.payload:
+            self._data = deserialize_data(self.payload[0])
+        return self._data
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, Message):
+            return NotImplemented
+        partial_comparison = (
+            self.version == other.version
+            and self.receiver == other.receiver
+            and self.sender == other.sender
+            and self.header == other.header
+        )
+        try:
+            my_data = self.data
+            other_data = other.data
+        except JSONDecodeError:
+            return partial_comparison and self.payload == other.payload
+        else:
+            return (partial_comparison and my_data == other_data
+                    and self.payload[1:] == other.payload[1:])
+
+    def __repr__(self) -> str:
+        return f"Message.from_frames({self._to_frames_without_sender_check()})"
+
+    # String access properties
+    @property
+    def receiver_str(self) -> str:
```

Comment:
those `*_str` accessors feel like avoidable API surface, but I don't have a better proposal ATM. Keep as strings, convert to bytes when outgoing/incoming? or keep as bytes, convert at (logging?) point of use (could be tedious)?

Reply:
In the Coordinator I like to do everything in bytes (no conversion needed). In user applications, I normally use strings, as they are more convenient for the user, who has to specify `inst_name = "Node.Actor"` and not the bytes version.

For that reason I added these str versions.

Reply:
Yeah, that might be a future enhancement to find a neater way to express that. Maybe there is none.